### PR TITLE
[ROCm] Unifying hip/cuda blas-lt APIs

### PR DESCRIPTION
**Pull request by pemeliya**
This is a follow-up PR for these two issues:

https://github.com/openxla/xla/pull/4406, https://github.com/openxla/xla/pull/3953

We unified hip/cuda blas-lt APIs by providing a common virtual interface defined in 
xla/stream_executor/gpu/gpu_blas_lt.h/.cc with implementations in 
xla/stream_executor/cuda/cuda_blas_lt.h/.cc and xla/stream_executor/rocm/hip_blas_lt.h/.cc, respectively.

The main design decision was that we made the class MatmulPlan (originally defined in xla/service/gpu/matmul_utils.h/.cc) **polymorphic** and moved it's interface declaration to gpu_blas_lt.h. 
There are two reasons for that, namely:

1. MatmulPlan provided a public function **ExecuteOnStream** which was implemented in terms of conditional compulation
with macros '#if GOOGLE_CUDA' or '#if TF_HIPBLASLT' in order to integrate library-specific data-types. This function becomes now a part of gpu_blas_lt interface.

2. MatmulPlan contained a library-specific member variable 'plan_' of type 'se::gpu::BlasLt::MatmulPlan' which is basically a plain container of MatmulDesc and several MatrixLayouts. These underlying types are again BLASLT library-specific and are **never** used directly, hence there is no need to expose BlasLt::MatmulDesc and BlasLt::MatrixLayout in the public interface.

Besides ExecuteOnStream, the class MatmulPlan also provides a number of overloaded 'DoMatmul' member functions (some of them are template functions) which were extracted as a common part from the original BlasLt implementations. These DoMatmul functions are also required for the oncoming integration of Blas-lt interface into Tensorflow: see tensorflow\core\kernels\matmul_util.h/.cc. 

We also extracted the library-specific argument type-checks from templated DoMatmul functions and moved them into a virtual function MatmulPlan::ValidateInputs().

The polymorphic class gpu::BlasLt (defined in gpu_blas_lt.h) is responsible for constructing the objects of type MatmulPlan, the rest blas-lt functionality is solely handled by MatmulPlan interface. 

The instantiations of gpu::BlasLt interface, as before, are defined in xla/stream_executor/cuda/cuda_blas.h and xla/stream_executor/rocm/rocm_blas.h, respectively.

We have also tried to compile the code with TF_HIPBLASLT=0 to make sure it also works fine if no hipblas-lt is available.

@akuegel: can you perhaps have a look at our implementation ?

### CODE REVIEW

**Comment by i-chaochen** [comment]
@ddunl could you find a related people to review this PR, please? Thanks!

File: xla/service/gpu/runtime/executable.h
Comment (pemeliya) [reply xla/service/gpu/runtime/executable.h:178]: ok, thanks, I will fix this

**Comment by i-chaochen** [comment]
Hi @akuegel wonder have you any suggestion on this PR? Thanks!

File: xla/service/gpu/gemm_rewriter.cc
Comment (akuegel) [inline xla/service/gpu/gemm_rewriter.cc:1849]: Why commenting this out?

File: xla/stream_executor/cuda/cuda_blas_lt.cc
Comment (akuegel) [inline xla/stream_executor/cuda/cuda_blas_lt.cc:53]: I think in general we prefer the separate lines for nested namespace declarations

File: xla/stream_executor/gpu/gpu_blas_lt.cc
Comment (akuegel) [inline xla/stream_executor/gpu/gpu_blas_lt.cc:1]: 2022 -> 2023

File: xla/stream_executor/gpu/gpu_blas_lt.h
Comment (akuegel) [inline xla/stream_executor/gpu/gpu_blas_lt.h:1]: 2022 -> 2023

**Review by akuegel — CHANGES_REQUESTED** [review]
General direction looks good :)

File: xla/service/gpu/gemm_rewriter.cc
Comment (pemeliya) [reply xla/service/gpu/gemm_rewriter.cc:1849]: This is basically the same change as was done by Wen in https://github.com/openxla/xla/pull/5852 in order to enable row-major  hipblas-lt support (to test my impl). But now as soon as PR 5852 is merged, I can rebase and remove this "fix".

**Comment by pemeliya** [comment]
rebased applied the recent reviewer changes

**Comment by radhakrishnaba** [comment]
Hi @akuegel , can you check this once? Thanks.

**Comment by gbaned** [comment]
Hi @pemeliya Can you please rebase your branch and resolve conflicts? Thank you!

File: xla/stream_executor/cuda/cuda_blas_lt.cc
Comment (akuegel) [inline xla/stream_executor/cuda/cuda_blas_lt.cc:40]: What about this TODO? I think the #if GOOGLE_CUDA might not be needed, as this code is guarded with if_cuda_is_configured.

File: xla/stream_executor/gpu/gpu_blas_lt.cc
Comment (akuegel) [inline xla/stream_executor/gpu/gpu_blas_lt.cc:94]: nit: please either use { } around this line, or move it to the previous line.
So 1-line if statements (with body) are ok, if it is a separate line, we prefer to use {} around the if body.

File: xla/stream_executor/cuda/cuda_blas_lt.cc
Comment (akuegel) [inline xla/stream_executor/cuda/cuda_blas_lt.cc:145]: I think it would be better to not leave this code in comments. Either delete, or uncomment. You could add a TODO comment instead if this needs a followup change. But usually just commenting out code looks like something left over from debugging and can confuse the reader of the code.

File: xla/stream_executor/cuda/cuda_blas_lt.cc
Comment (akuegel) [inline xla/stream_executor/cuda/cuda_blas_lt.cc:161]: Same here, please either delete this code (plus possibly add a TODO), or uncomment

File: xla/stream_executor/gpu/gpu_blas_lt.cc
Comment (akuegel) [inline xla/stream_executor/gpu/gpu_blas_lt.cc:104]: nit: Please everything in one line, or use {} around the if body.

File: xla/stream_executor/cuda/cuda_blas_lt.cc
Comment (pemeliya) [reply xla/stream_executor/cuda/cuda_blas_lt.cc:40]: yes, you are right, I was confused by ROCm side where we have an additional macro TF_HIPBLASLT which might be set to 0 even if ROCm is configured. I will remote this

File: xla/stream_executor/cuda/cuda_blas_lt.cc
Comment (pemeliya) [reply xla/stream_executor/cuda/cuda_blas_lt.cc:145]: yes, I will remove it: this is a letfover after I changed the prototype of the function BlasLt::MatrixLayout::Create(). Previously, loading_dim_stride and batch_stride were std::optional<>, now they are always set as members of gpu::MatrixLayout datatype, hence this "intelligence" is no longer needed..

File: xla/stream_executor/gpu/gpu_blas_lt.cc
Comment (pemeliya) [reply xla/stream_executor/gpu/gpu_blas_lt.cc:104]: ok, I see, I will try to find other similar places in my code, if any

**Comment by i-chaochen** [comment]
@akuegel the feedback is resolved, could you have another look, please. Thanks

**Review by akuegel — APPROVED** [review]

**Comment by akuegel** [comment]
It seems there are also adjustments needed for tensorflow (matmul_util.h, matmul_util.cc and matmul_op_impl.h in core/kernels directory). I am not sure what the proper process is for a cross-cutting change like this. But maybe you can create a separate PR for tensorflow and I will copy it from there before merging? Internally, we need to do this as one change.

**Comment by pemeliya** [comment]
@akuegel I have prepared a separate Tensorflow PR for integration of new blas-lt API:
https://github.com/tensorflow/tensorflow/pull/62099
can you please have a look ? 
I also had to apply several small changes to this XLA branch to make the interface compatible.

**Review by akuegel — APPROVED** [review]

**Comment by wenchenvincent** [comment]
@akuegel Thanks for approving the PR! It seems that it still shows that it needs an approving review. Do we need to get approval from another reviewer?

**Comment by akuegel** [comment]
My guess is that this "Review required" refers to the Gerrit review that we now actually skip, but it is still being requested if a PR has additional changes and needs another approval. In fact, that review is not required, and approving here directly imports those changes into the internal change.
I didn't have enough time yesterday to merge the tensorflow PR with the additional fixes into the internal change, will work on that now.

**Pull request closed without merging**
